dynamic_programming/longest_common_string.cpp

```diff
@@ -1,53 +1,142 @@
+};
+
+std::vector<TestCase> get_test_cases() {
+    return {TestCase("", "", 0),
+            TestCase("ab", "ab", 2),
+            TestCase("ab", "ba", 1),
+            TestCase("", "xyz", 0),
+            TestCase("abcde", "ace", 3),
+            TestCase("BADANA", "ANADA", 3),
+            TestCase("BADANA", "CANADAS", 3),
+            TestCase("a1a234a5aaaa6", "A1AAAA234AAA56AAAAA", 6),
+            TestCase("123x", "123", 3),
+            TestCase("12x3x", "123", 3),
+            TestCase("1x2x3x", "123", 3),
+            TestCase("x1x2x3x", "123", 3),
+            TestCase("x12x3x", "123", 3)};
+}
+
+/**
+ * @brief checks the function longest_common_string_length agains example data
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length(const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(cur_tc.str_a, cur_tc.str_b) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief checks the function longest_common_string_length returns the same
+ * result when its argument are flipped
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length_is_symmetric(
+    const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(cur_tc.str_b, cur_tc.str_a) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief reverses a given string
+ * @param in_str input string
+ * @return the string in which the characters appear in the reversed order as in
+ * in_str
+ */
+std::string reverse_str(const std::string& in_str) {
+    return {in_str.rbegin(), in_str.rend()};
+}
 
-    cout << "longest string size = " << ma /*<<" "<<indi<<" "<<indj*/ << "\n";
-    for (i = indi - 3; i < indi; i++) cout << str1[i];
-    cout << "\n";
+/**
+ * @brief checks the function longest_common_string_length returns the same
+ * result when its inputs are reversed
+ * @param test_cases list of test cases
+ * @tparam type representing a list of test cases
+ */
+template <typename TestCases>
+void test_longest_common_string_length_for_reversed_inputs(
+    const TestCases& test_cases) {
+    for (const auto& cur_tc : test_cases) {
+        assert(longest_common_string_length(reverse_str(cur_tc.str_a),
+                                            reverse_str(cur_tc.str_b)) ==
+               cur_tc.common_string_len);
+    }
+}
+
+/**
+ * @brief runs all tests for longest_common_string_length funcion
+ */
+void tests() {
+    const auto test_cases = get_test_cases();
+    assert(test_cases.size() > 0);
+    test_longest_common_string_length(test_cases);
+    test_longest_common_string_length_is_symmetric(test_cases);
+    test_longest_common_string_length_for_reversed_inputs(test_cases);
```

Comment:
I'm sorry but i feel as if the testing is unnecessarily complex and i cant even understand whats going on could you reduce it down to just assert statements? or explain the complexities
you could possibly just use comments to organize your tests instead of make whole functions for each type. i feel as if a beginner would struggle at understanding how any of these tests work

Reply:
One could rename these function to:
```cpp
test_if_longest_common_string_length_returns_the_correct_result(test_cases);
test_if_longest_common_string_length_returns_the_correct_result_when_its_arguments_are_swapped(test_cases);
test_if_longest_common_string_length_returns_the_correct_results_when_its_inputs_are_reversed(test_cases);
```
but I think these names are slightly too long.

These tests also _check_ few more things then the previous ones and they _scale_ (i.e. it is easy to add a new test case, when needed).

Reply:
the function names are not the issue the variable names are

Reply:
> the function names are not the issue the variable names

Which variable names are the issue?

> tests are for simply testing as long as you have tests showcasing most of the possibilities that would be enough no?

No! Person writing a test should always try to **break** the code!

Especially in case of such _puzzle solving_, it is important to invest some more resources in tests. It might happen, that in some future PR, someone will make some very aggressive optimization of the solution, which might be incorrect. Thanks to _good_ tests, such mistake can be spotted.

@realstealthninja can I mark it as _resolved_?

Reply:
Yes I understand the concern and i think i agree with you but such drastic measures are bit overboard no? One could just add multiple asserts to see the various cases like for example an algorithm which takes two numbers and adds divides could produce negative and  postive numbers there is also a case for an error on division with zero these could all be represented via simple assert statements.

Besides drastic changes of code go through 2 repository maintainers and the code is usually required to be more readable than faster. 

Speed is not the goal of this repository it's readability and ingenuity coming together to form an understand in a beginner.

My main concern is I've seen some algorithms reinventing the assert statement with define statements.
People are more likely to contribute by looking at previous contributions and not the contribution guidelines and this sets a precedent of making testing easier on the cost of readability

IMO tests should be just assert statements unless they physically cannot be.
but i would love to hear your thoughts on this

Reply:
Usually it is not a good idea to reinvent the wheel. However, using just the simple `assert`'s might be sometime cumbersome (like in the case discussed right now - almost every _test case_ would need to be copied into three places).

[I highly suggest using some testing framework](https://github.com/TheAlgorithms/C-Plus-Plus/pull/2235#issuecomment-1279699322). Then one could simply express these tests as _parametrised tests_ (cf. [GoogleTest](http://google.github.io/googletest/advanced.html#value-parameterized-tests), [boost.test](https://www.boost.org/doc/libs/1_82_0/libs/test/doc/html/boost_test/tests_organization/test_cases/test_case_generation.html#boost_test.tests_organization.test_cases.test_case_generation.data_driven_tests_in_the_boost_t) or [Catch2](https://github.com/catchorg/Catch2/blob/devel/docs/generators.md)). In general I agree that the logic of the test should be simple - the complexity should be handled by the framework.

Reviewers are people and make mistakes (how did the original file ended here?).

So... can I mark it as resolved?

Reply:
Yes mark this as solved! Thank you for educating me on this!!!